cli/lib/kontena/cli/stacks/yaml/stack_file_loader.rb

```diff
@@ -0,0 +1,96 @@
+require_relative '../stack_name'
+require 'yaml'
+require_relative 'reader'
+
+module Kontena::Cli::Stacks
+  module YAML
+    class StackFileLoader
+      # A base class for loading stack files. You can define more loaders by
+      # inheriting from this class.
+      #
+      # The purpose of StackFileLoader is to provide a generic interface for
+      # loading stack YAML's from different sources, such as local files,
+      # stack registry or URLs
+
+      def self.inherited(where)
+        loaders << where
+      end
+
+      def self.loaders
+        @loaders ||= []
+      end
+
+      # The main interface for getting a new loader
+      #
+      # @param source [String] stack file source string (filename, url, ..)
+      # @param parent [StackFileLoader] define a parent for recursion
+      # @return [StackFileLoader]
+      def self.for(source, parent = nil)
+        loader = loaders.find { |l| l.match?(source, parent) }
+        raise "Can't determine stack file origin for '#{source}'" if loader.nil?
+        loader.new(source, parent)
+      end
+
+      attr_reader :source, :parent
+
+      # @param source [String] stack file source string (filename, url, ..)
+      # @param parent [StackFileLoader] define a parent for recursion
+      # @return [StackFileLoader]
+      def initialize(source, parent = nil)
+        @source = source
+        @parent = parent
+        set_context if respond_to?(:set_context)
```

Comment:
This took a little bit of hunting to figure out what's going on here... have the `FileLoader` override `initialize instead`?

```ruby
def initialize(source, parent = nil)
  super(self.class.with_context(source, parent), parent)
end
```